File: frontend/invest/src/types/lossCutApproval.ts
export type EvidenceStatus =
  | "filled"
  | "stale"
  | "missing"
  | "unavailable"
  | "source_error";

export interface LossCutEvidenceField {
  status: EvidenceStatus;
  label: string;
  value: Record<string, unknown> | null;
  reason: string | null;
  source: string | null;
  as_of: string | null;
  valid_until: string | null;
}

export interface LossCutPositionEvidence {
  account_ref: string;
  account_mode: string;
  market: string;
  symbol: string;
  total_quantity: string;
  sellable_quantity: string | null;
  pending_sell_quantity: string | null;
  average_price: string | null;
  current_price: string | null;
  source: string;
  source_status: EvidenceStatus;
  source_reason: string | null;
  observed_at: string;
}

export interface LossCutEvidenceResponse {
  mode: "symbol" | "proposal";
  symbol: string;
  proposal_id: string | null;
  generated_at: string;
  can_begin: boolean;
  positions: LossCutPositionEvidence[];
  loss: LossCutEvidenceField;
  reason: LossCutEvidenceField;
  r931: LossCutEvidenceField;
  consensus: LossCutEvidenceField;
  watch: LossCutEvidenceField;
  fingerprint: Record<string, unknown> | null;
  warnings: string[];
}

export interface LossCutBeginResponse {
  proposal_id: string;
  ceremony_id: string;
  expires_at: string;
  evidence: LossCutEvidenceResponse;
  fingerprint: Record<string, unknown>;
  next_step: "confirm";
}

export interface LossCutConfirmResponse {
  proposal_id: string;
  status: "validated_no_execution";
  evidence: LossCutEvidenceResponse;
  fingerprint: Record<string, unknown>;
}

File: app/schemas/loss_cut_approval.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

EvidenceStatus = Literal["filled", "stale", "missing", "unavailable", "source_error"]


class LossCutEvidenceField(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: EvidenceStatus
    label: str
    value: dict[str, Any] | None = None
    reason: str | None = None
    source: str | None = None
    as_of: str | None = None
    valid_until: str | None = None


class LossCutPositionEvidence(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_ref: str
    account_mode: str
    market: str
    symbol: str
    total_quantity: str
    sellable_quantity: str | None
    pending_sell_quantity: str | None
    average_price: str | None
    current_price: str | None
    source: str
    source_status: EvidenceStatus
    source_reason: str | None = None
    observed_at: str


class LossCutEvidenceResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mode: Literal["symbol", "proposal"]
    symbol: str
    proposal_id: str | None = None
    generated_at: str
    can_begin: bool
    positions: list[LossCutPositionEvidence] = Field(default_factory=list)
    loss: LossCutEvidenceField
    reason: LossCutEvidenceField
    r931: LossCutEvidenceField
    consensus: LossCutEvidenceField
    watch: LossCutEvidenceField
    fingerprint: dict[str, Any] | None = None
    warnings: list[str] = Field(default_factory=list)


class LossCutBeginRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")


class LossCutBeginResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    proposal_id: str
    ceremony_id: str
    expires_at: str
    evidence: LossCutEvidenceResponse
    fingerprint: dict[str, Any]
    next_step: Literal["confirm"] = "confirm"


class LossCutConfirmRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    ceremony_id: str = Field(min_length=32, max_length=256)


class LossCutConfirmResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    proposal_id: str
    status: Literal["validated_no_execution"] = "validated_no_execution"
    evidence: LossCutEvidenceResponse
    fingerprint: dict[str, Any]
